### ISSUE TITLE
Automated checking of input and output?

### ISSUE THREAD

**Issue by tpilz** [echse/echse_generic#2]
At the moment I am working on a rather large and bulky model engine and have some problems with bugs in my code producing NaN values etc. which are not easy to detect. I try to catch this with the isnan macro defined in the cmath library. Furthermore, I defined a NA value (e.g. -9999) as shared parameter and assign that value to parameters I currently do not need within a specific model setup (e.g. when selecting among different process realisations). However, I manually included checks in function definitions to ensure that all currently needed parameters are not NA (as, by definition, all declared parameters, inputs etc. have always to be defined in ECHSE).

I wonder if it is possible to enhance the core in a way that checking of input (including parameters etc.) and output values (e.g. for a user-defined NA-value or if values are finite and not NaN etc.) is generalized and does not has to be done explicitly within the user-specific code of the model engine. I think that would greatly simplify model building for a user.

Cheers,

Tobias

PS: I wanted to assign an enhancement label to that issue but I can't. Not sure if this can only be done by the owner of the repo?! In my own repo I can do it when creating an issue.

**Comment by echse** [comment]
Hi Tobias,

I think one needs to distinguish the two cases:

(1) invalid floating point numbers

(2) special values defined by the user (being valid numbers)

Infos for case 1:

There is an ECHSE configuration parameter "trap_fpe" (page 39 in 
echse_core_doc.pdf). If set to "true", the states and outputs of every 
object are checked for invalid floating point numbers, essentially using 
the C++ function "isfinite" (e. g. 
http://www.cplusplus.com/reference/cmath/isfinite). IMO this function is 
often more useful than "isnan" as it detects also other types of problems.

Like isnan, you could use isfinite within your own class code for the 
purpose of debugging.

Infos for case 2:

Not sure whether checks for user-defined special values should be 
implemented at the ECHSE's core level. This might cost performance even 
if no such checks are actually required.

Furthermore, those special values (like the common -9999) are not 
universally applicable. In some context, -9999 might be a legal value 
such as a spatial coordinate or water depth. Hence, different special 
values might be required for different parameters.

What is the actual problem with your current approach?

I think I would write my own function "is_special" that takes a double 
value (or a vector of doubles) as input and compares the value(s) 
against the preferred special constant. Through a header file, the 
function can be made available to several class codes, if necessary. 
Doesn't that work?

David

**Comment by tpilz** [comment]
Hi David,

thanks for your fast response!

Case 1: 
**EDIT (ignore quoted text below)**

> Right, I forgot about 'trap_fpe'. In fact, my specific problem was not an issue of NaN as I just recognised. For an output variable I got a value of '-nan' but neither the echse built-in check threw an error nor explicitly testing the variable with isfinite() or isnan() returned true. However, I discovered that it was just a very very small number I could print during runtime (1.92747e-22) but the output (in this case within a *.dbg file) gave '-nan'. So this was rather a precision problem misleading me into a wrong direction.

Sorry, the above is wrong (in case you've read it before my edit). True is:
A function contains a bug and produces a NaN value which is then used for further calculations (producing further NaN) which are then written as output. However, the NaNs are not recognised, neither by the ECHSE core nor by isnan(). I don't really understand what is going on here. If I test the erroneous function separately (in a separate program outside of echse) the NaNs are correctly recognised by inan(). Maybe the problem is just my limited knowledge of C++. However, I try to figure out what is happening ...

Case 2:

I know it is problematic to universally define a value for NA. Although I thought it might be reasonable to let the user choose a global value for a specific setup (where the user can estimate which is an applicable value). My motivation is more or less a mixture of laziness and readability of source code. 

Example: I have several functions for evapotranspiration. All these take into account different inputs and variables. Which function is used can be defined by a shared parameter. However, for the pre-processing I have to define all possible parameters and possibly assign a dummy value if they are not needed in the current setup. I defined a shared parameter as NA value and when calling a specific function ensure that all needed parameters are not NA. That means for every parameter I have to write something like this within the function code:

``` c++
if ( abs(eddy_decay - na_val) < 0.01 ) {
    stringstream errmsg;
    errmsg << "Eddy diffusivity decay constant is missing!";
    except e(__PRETTY_FUNCTION__,errmsg,__FILE__,__LINE__);
    throw(e);
}
```

That is somewhat annoying and reduced readability of the source code so I was thinking about something more efficient. You are right, writing a function like "is_special" would at least reduce the problem a bit but still requires some lines of code (for complex function taking a lot of arguments). 

Maybe there is a better solution? Although I acknowledge it might be hard (or impossible) to realise that within the ECHSE core.

Tobias

**Comment by echse** [comment]
Hi Tobias,

I updated the repository; specifically I added a member function 
"checkParamNum" in the file "echse_coreClass_abstractObject.h". Please 
have a look it.

Within your simulate method (of any class) you can now call this method 
like so:

checkParamNum( theNameConstant, NAvalue, tolerance);

for example

checkParamNum( soildepth, -9999, 0.001);

If the check fails, the method throws an exception (being caught 
automatically, if you don't catch it explicitly).

Please check and give me feedback on the question of whether this should 
become a permanent feature.

David

**Comment by tpilz** [comment]
Hi all,

the issues are solved.

Case 1: 

This was caused by a bug in the compilation file, see commit d243b392c1e1d9bf57d25e67eb673e436f0f0cdb

Case 2: 

Partly solved by commit 5879f15f3af4051866aab489d77ab4671dc67f5e

The methods checkParamNum() and checkSharedParamNum(), however, can only be used within the simulate method to check parameters. For my individual problem mentioned above the best solution probably is to use David's suggestion and define a check function (similar to checkParamNum()) in a header file which can then be used to check function arguments. This slightly reduces the amount of code (in contrast to my example above).

Cheers,
Tobias
